From 395728c607e2bbac32ef84d89705716133c4fb50 Mon Sep 17 00:00:00 2001
From: Mikita Lipski <mikita.lipski@amd.com>
Date: Mon, 9 Apr 2018 13:40:00 -0400
Subject: [PATCH 4757/5725] drm/amd/display: Apply clock for voltage request

Translate dm_pp tructure to pp type
Call PP lib to apply clock voltage request for display

Signed-off-by: Mikita Lipski <mikita.lipski@amd.com>
Reviewed-by: Tony Cheng <Tony.Cheng@amd.com>
Acked-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
---
 .../drm/amd/display/amdgpu_dm/amdgpu_dm_services.c | 31 ++++++++++++++++++++--
 1 file changed, 29 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_services.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_services.c
index 9e923a4..985e69e 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_services.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_services.c
@@ -375,8 +375,35 @@ bool dm_pp_apply_clock_for_voltage_request(
 	const struct dc_context *ctx,
 	struct dm_pp_clock_for_voltage_req *clock_for_voltage_req)
 {
-	/* TODO: to be implemented */
-	return false;
+	struct amdgpu_device *adev = ctx->driver_context;
+	struct pp_display_clock_request *pp_clock_request = {0};
+	int ret = 0;
+	switch (clock_for_voltage_req->clk_type) {
+	case DM_PP_CLOCK_TYPE_DISPLAY_CLK:
+		pp_clock_request->clock_type = amd_pp_disp_clock;
+		break;
+
+	case DM_PP_CLOCK_TYPE_DCEFCLK:
+		pp_clock_request->clock_type = amd_pp_dcef_clock;
+		break;
+
+	case DM_PP_CLOCK_TYPE_PIXELCLK:
+		pp_clock_request->clock_type = amd_pp_pixel_clock;
+		break;
+
+	default:
+		return false;
+	}
+
+	pp_clock_request->clock_freq_in_khz = clock_for_voltage_req->clocks_in_khz;
+
+	if (adev->powerplay.pp_funcs->display_clock_voltage_request)
+		ret = adev->powerplay.pp_funcs->display_clock_voltage_request(
+			adev->powerplay.pp_handle,
+			pp_clock_request);
+	if (ret)
+		return false;
+	return true;
 }
 
 bool dm_pp_get_static_clocks(
-- 
2.7.4

